src/commands/BootstrapCommand.js

```diff
@@ -133,32 +134,251 @@ export default class BootstrapCommand extends Command {
+      return pkg.name === name && (!version || semver.satisfies(pkg.version, version));
+    });
+
+    const hasPackage = (name, version) => Boolean(findPackage(name, version));
+
+    // Configuration for what packages to hoist may be in lerna.json or it may
+    // come in as command line options.
+    const {hoist: scope, nohoist: ignore} = this.getOptions();
+
+    // This will contain entries for each hoistable dependency.
+    const root = [];
+
+    // This will map packages to lists of unhoistable dependencies
+    const leaves = {};
+
+    /**
+     * Map of dependencies to install
+     * {
+     *   <name>: {
+     *     versions: {
+     *       <version>: <# of dependents>
+     *     },
+     *     dependents: {
+     *       <version>: [<dependent1>, <dependent2>, ...]
+     *     }
+     *   }
+     * }
+     *
+     * Example:
+     *
+     * {
+     *   react: {
+     *     versions: {
+     *       "15.x": 3,
+     *       "^0.14.0": 1
+     *     },
+     *     dependents: {
+     *       "15.x": ["my-component1", "my-component2", "my-component3"],
+     *       "^0.14.0": ["my-component4"],
+     *     }
+     *   }
+     * }
+     */
+    const depsToInstall = {};
+
+    // get the map of external dependencies to install
+    packages.forEach((pkg) => {
+
+      // for all package dependencies
+      Object.keys(pkg.allDependencies)
+
+        // map to package or normalized external dependency
+        .map((name) => findPackage(name, pkg.allDependencies[name]) || { name, version: pkg.allDependencies[name] })
+
+        // match external and version mismatched local packages
+        .filter((dep) => !hasPackage(dep.name, dep.version) || !pkg.hasMatchingDependency(dep))
+
+        .forEach(({name, version}) => {
+
+          // Get the object for this package, auto-vivifying.
+          const dep = depsToInstall[name] || (depsToInstall[name] = {
+            versions   : {},
+            dependents : {}
+          });
+
+          // Add this version if it's the first time we've seen it.
+          if (!dep.versions[version]) {
+            dep.versions  [version] = 0;
+            dep.dependents[version] = [];
+          }
+
+          // Record the dependency on this version.
+          dep.versions  [version]++;
+          dep.dependents[version].push(pkg.name);
+        });
+    });
+
+    // determine where each dependency will be installed
+    Object.keys(depsToInstall).forEach((name) => {
+      const {versions, dependents} = depsToInstall[name];
+
+      let rootVersion;
+
+      if (scope && PackageUtilities.getFilteredPackage({name}, {scope, ignore})) {
+
+        // Get the most common version.
+        const commonVersion = Object.keys(versions)
+          .reduce((a, b) => versions[a] > versions[b] ? a : b);
+
+        // Get the version required by the repo root (if any).
+        // If the root doesn't have a dependency on this package then we'll
+        // install the most common dependency there.
```

Comment:
@evocateur i experimented with trying to get commands such as `npm outdated` to work by symlinking the hoisted deps at the root to the `node_modules` of each package. this caused lots of issues with the way npm treats symlinks. IMO, it wasn't worth it.

one of the other problems i ran into was forgetting to add a dependency to a package's `package.json` file because it was hoisted from another package and everything worked as expected. i thought of creating a tool for checking these sorts of things, but never got around to it.